File: leetcode/valid_palindrome.py
# Given a string, determine if it is a palindrome, considering only 
# alphanumeric characters and ignoring cases.
# Time Complexity: O(n)
# Space Complexity: O(1)

class Solution:
    # @param {string} s
    # @return {boolean}
    def isPalindrome(self, s):
        left, right = 0, len(s) - 1
        while left < right:
            while left < right and not s[left].isalnum():
                left += 1
            while right > left and not s[right].isalnum():
                right -= 1
            if s[left].lower() != s[right].lower():
                return False
            left += 1
            right -= 1
        return True

if __name__ == "__main__":
    solution = Solution()
    print solution.isPalindrome("A man, a plan, a canal: Panama")
    print solution.isPalindrome("race a car")

File: leetcode/valid_palindrome.js
/* Given a string, determine if it is a palindrome, considering only 
 * alphanumeric characters and ignoring cases.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
    for (var left = 0, right = s.length - 1; left < right; ++left, --right) {
        while (left < right && !/^[a-z0-9]+$/i.test(s[left]))
            ++left
        while (left < right && !/^[a-z0-9]+$/i.test(s[right]))
            --right
        if (s[left].toLowerCase() !== s[right].toLowerCase())
            return false
    }    
    return true
};
